pkg/distributor/distributor.go

```diff
@@ -1042,11 +1045,33 @@ func (d *Distributor) prePushHaDedupeMiddleware(next PushFunc) PushFunc {
 		numSamples := 0
 		now := time.Now()
 		group := d.activeGroups.UpdateActiveGroupTimestamp(userID, validation.GroupLabel(d.limits, userID, req.Timeseries), now)
+		sampleTimestamp := timestamp.FromTime(now)
+		if d.limits.HATrackerUseSampleTimeForFailover(userID) {
+			earliestSampleTimestamp := sampleTimestamp
+			latestSampleTimestamp := sampleTimestamp
+			for _, ts := range req.Timeseries {
+				if len(ts.Samples) > 0 {
+					tsms := ts.Samples[0].TimestampMs
+					if tsms < earliestSampleTimestamp {
+						earliestSampleTimestamp = tsms
+					} else if tsms > latestSampleTimestamp {
+						latestSampleTimestamp = tsms
+					}
+				}
+			}
+			delta := now.UnixMilli() - earliestSampleTimestamp
+			if delta > 10000 {
+				delta2 := sampleTimestamp - earliestSampleTimestamp
+				delta3 := latestSampleTimestamp - now.UnixMilli()
+				level.Warn(d.log).Log("msg", "earliest sample timestamp is a lot earlier than now", "delta", delta, "delta2", delta2, "delta3", delta3, "user", userID, "cluster", cluster, "replica", replica, "earliestSampleTime", timestamp.Time(earliestSampleTimestamp), "sampleTime", timestamp.Time(sampleTimestamp), "latestSampleTime", timestamp.Time(latestSampleTimestamp), "now", now)
```

Comment:
can delta1, delta2, delta3 have better names?

Reply:
This logging is just temporary, for me to look at. If we keep this approach, I'll probably simplify these logs (get rid of the extra deltas) or remove them altogether, but if/when we start trying to merge to main, I'll make sure to check that the names of the stuff are more descriptive. I've renamed them, but since the only changes so far are to the logging, docs and tests without changing any functionality, I won't build and test a new image (unless we decide we need to change any functionality here)